nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java

```diff
@@ -0,0 +1,151 @@
+      NessieClient client) {
+    this.conf = conf;
+    this.key = key;
+    this.reference = reference;
+    this.client = client;
+  }
+
+  @Override
+  protected void doRefresh() {
+    // break reference with parent (to avoid cross-over refresh)
+    // TODO, confirm this is correct behavior.
+    // reference = reference.copy();
+
+    reference.refresh();
+    String metadataLocation = null;
+    try {
+      Contents contents = client.getContentsApi().getContents(key, reference.getHash());
+      this.table = contents.unwrap(IcebergTable.class)
+          .orElseThrow(() ->
+              new IllegalStateException("Cannot refresh iceberg table: " +
+                  String.format("Nessie points to a non-Iceberg object for path: %s.", key)));
+      metadataLocation = table.getMetadataLocation();
+    } catch (NessieNotFoundException ex) {
+      if (currentMetadataLocation() != null) {
+        throw new NoSuchTableException(ex, "No such table %s", key);
+      }
+    }
+    refreshFromMetadataLocation(metadataLocation, 2);
+  }
+
+  @Override
+  protected void doCommit(TableMetadata base, TableMetadata metadata) {
+    reference.checkMutable();
+
+    String newMetadataLocation = writeNewMetadata(metadata, currentVersion() + 1);
+
+    try {
+      IcebergTable newTable = ImmutableIcebergTable.builder().metadataLocation(newMetadataLocation).build();
+      client.getContentsApi().setContents(key,
+                                          reference.getAsBranch().getName(),
+                                          reference.getHash(),
+                                          String.format("iceberg commit%s", applicationId()),
+                                          newTable);
+    } catch (NessieConflictException ex) {
+      io().deleteFile(newMetadataLocation);
+      String fixMsg = reference.isBranch() ?
+          String.format("Update the reference %s and try again", reference.getName()) :
+          String.format("Can't commit to the tag %s", reference.getName());
+      throw new CommitFailedException(ex, "Commit failed: Reference hash is out of date. %s", fixMsg);
+    } catch (NessieNotFoundException ex) {
+      io().deleteFile(newMetadataLocation);
+      throw new RuntimeException(String.format("Commit failed: Reference %s does not exist", reference.getName()), ex);
+    } catch (Throwable e) {
+      io().deleteFile(newMetadataLocation);
+      throw new RuntimeException("Unexpected commit exception", e);
+    }
+  }
+
+  @Override
+  public FileIO io() {
+    if (fileIO == null) {
+      fileIO = new HadoopFileIO(conf);
+    }
+
+    return fileIO;
+  }
+
+  /**
+   * try and get a Spark application id if one exists.
+   *
+   * <p>
+   *   We haven't figured out a general way to pass commit messages through to the Nessie committer yet.
+   *   This is hacky but gets the job done until we can have a more complete commit/audit log.
+   * </p>
+   */
+  private static String applicationId() {
```

Comment:
For Spark writes, we pass the application ID in through `snapshot.summary()`: https://github.com/apache/iceberg/blob/9af545ed56343b2fa09966166f8da3e7a24100d7/spark3/src/main/java/org/apache/iceberg/spark/source/SparkBatchWrite.java#L153

Reply:
Ahh, thanks for that. Much cleaner this way.